src/main/java/com/mojang/serialization/Codec.java

```diff
@@ -0,0 +1,623 @@
+                return decoder.decode(ops, input);
+            }
+
+            @Override
+            public <T> RecordBuilder<T> encode(final A input, final DynamicOps<T> ops, final RecordBuilder<T> prefix) {
+                return encoder.encode(input, ops, prefix);
+            }
+
+            @Override
+            public String toString() {
+                return name;
+            }
+        };
+    }
+
+    static <F, S> Codec<Pair<F, S>> pair(final Codec<F> first, final Codec<S> second) {
+        return new PairCodec<>(first, second);
+    }
+
+    static <F, S> Codec<Either<F, S>> either(final Codec<F> first, final Codec<S> second) {
+        return new EitherCodec<>(first, second);
+    }
+
+    static <F, S> MapCodec<Pair<F, S>> mapPair(final MapCodec<F> first, final MapCodec<S> second) {
+        return new PairMapCodec<>(first, second);
+    }
+
+    static <F, S> MapCodec<Either<F, S>> mapEither(final MapCodec<F> first, final MapCodec<S> second) {
+        return new EitherMapCodec<>(first, second);
+    }
+
+    static <E> Codec<List<E>> list(final Codec<E> elementCodec) {
+        return new ListCodec<>(elementCodec);
+    }
+
+    static <K, V> Codec<List<Pair<K, V>>> compoundList(final Codec<K> keyCodec, final Codec<V> elementCodec) {
+        return new CompoundListCodec<>(keyCodec, elementCodec);
+    }
+
+    static <K, V> SimpleMapCodec<K, V> simpleMap(final Codec<K> keyCodec, final Codec<V> elementCodec, final Keyable keys) {
+        return new SimpleMapCodec<>(keyCodec, elementCodec, keys);
+    }
+
+    static <K, V> UnboundedMapCodec<K, V> unboundedMap(final Codec<K> keyCodec, final Codec<V> elementCodec) {
+        return new UnboundedMapCodec<>(keyCodec, elementCodec);
+    }
+
+    static <F> MapCodec<Optional<F>> optionalField(final String name, final Codec<F> elementCodec) {
+        return new OptionalFieldCodec<>(name, elementCodec);
+    }
+
+    default Codec<List<A>> listOf() {
+        return list(this);
+    }
+
+    default <S> Codec<S> xmap(final Function<? super A, ? extends S> to, final Function<? super S, ? extends A> from) {
+        return Codec.of(comap(from), map(to), toString() + "[xmapped]");
+    }
+
+    default <S> Codec<S> comapFlatMap(final Function<? super A, ? extends DataResult<? extends S>> to, final Function<? super S, ? extends A> from) {
+        return Codec.of(comap(from), flatMap(to), toString() + "[comapFlatMapped]");
+    }
+
+    default <S> Codec<S> flatComapMap(final Function<? super A, ? extends S> to, final Function<? super S, ? extends DataResult<? extends A>> from) {
+        return Codec.of(flatComap(from), map(to), toString() + "[flatComapMapped]");
+    }
+
+    default <S> Codec<S> flatXmap(final Function<? super A, ? extends DataResult<? extends S>> to, final Function<? super S, ? extends DataResult<? extends A>> from) {
+        return Codec.of(flatComap(from), flatMap(to), toString() + "[flatXmapped]");
+    }
```

Comment:
Proposal: if you introduce some helper interface like `FlatFunction<T, R> extends Function<T, DataResult<R>>` you probably should be able to use overrides instead of those weird names.

Reply:
But then calling with lambdas will break on anbiguous method call :)